tag.go

```diff
@@ -0,0 +1,135 @@
+package git
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"gopkg.in/src-d/go-git.v3/core"
+)
+
+// Tag represents an annotated tag object. It points to a single git object of
+// any type, but tags typically are applied to commit or blob objects. It
+// provides a reference that associates the target with a tag name. It also
+// contains meta-information about the tag, including the tagger, tag date and
+// message.
+//
+// https://git-scm.com/book/en/v2/Git-Internals-Git-References#Tags
+type Tag struct {
+	Hash    core.Hash
+	Type    core.ObjectType
+	Name    string
+	Tagger  Signature
+	Message string
+
+	object core.Hash
+	r      *Repository
+}
+
+// Decode transforms a core.Object into a Tag struct.
+func (t *Tag) Decode(o core.Object) error {
+	if o.Type() != core.TagObject {
+		return ErrUnsupportedObject
+	}
+
+	t.Hash = o.Hash()
+
+	r := bufio.NewReader(o.Reader())
+	for {
+		line, err := r.ReadSlice('\n')
+		if err != nil && err != io.EOF {
+			return err
+		}
+
+		line = bytes.TrimSpace(line)
+		if len(line) == 0 {
+			break // Start of message
+		}
+
+		split := bytes.SplitN(line, []byte{' '}, 2)
+		switch string(split[0]) {
+		case "object":
+			t.object = core.NewHash(string(split[1]))
+		case "type":
+			t.Type, err = core.ParseObjectType(string(split[1]))
+			if err != nil {
+				return err
+			}
+		case "tag":
+			t.Name = string(split[1])
+		case "tagger":
+			t.Tagger.Decode(split[1])
+		}
+
+		if err == io.EOF {
+			return nil
+		}
+	}
+
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		return err
+	}
+	t.Message = string(data)
+
+	return nil
+}
+
+// Commit returns the commit pointed to by the tag. If the tag points to a
+// different type of object an error will be returned.
+func (t *Tag) Commit() (*Commit, error) {
+	return t.r.Commit(t.object)
+}
+
+// Tree returns the tree pointed to by the tag. If the tag points to a
+// different type of object an error will be returned.
+func (t *Tag) Tree() (*Tree, error) {
+	// TODO: If the tag is of type commit, follow the commit to its tree?
+	return t.r.Tree(t.object)
+}
+
+// TODO: Add support for retrieving blobs? We don't have a type for that
+//       currently.
```

Comment:
`Tag` does have an `Object()` function which could be used to read blob data via `Object.Reader()`, but that won't be obvious to new users without an example.

I think we should at least add a `Tag.Blob()` function that only returns a value if the object's type is actually `BlobObject`. I think that's worthwhile even if it just returns a `core.Object` or a new `core.Blob` type alias. It could be used like so:

```
func processTag(tag Tag) {
	switch tag.Type {
	case core.CommitObject:
		commit, err := tag.Commit()
		// Do something with the commit
	case core.BlobObject:
		blob, err := tag.Blob()
		// Do something with the blob
	default:
		// Something we aren't interested in
	}
}
```

I'll go ahead and add the `Tag.Blob()` function to this PR. Giving a blob a friendly name and message is a legitimate use case for annotated tags, so I think it's important to support that scenario.

We could go so far as to create a new `Blob` struct along with a `BlobIter`, but if we do I think we should put that in a separate PR.